src/main/java/org/opensearch/knn/index/KNNWeight.java

```diff
@@ -117,12 +124,34 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
             final KNNQueryResult[] results;
             KNNCounter.GRAPH_QUERY_REQUESTS.increment();
 
+            // We need to first get index allocation
+            NativeMemoryAllocation indexAllocation = null;
+
+            try {
+                indexAllocation = nativeMemoryCacheManager.get(
```

Comment:
What happens if we do not get the allocation because cache is full or the index is more than cache size?

Reply:
So, I believe the cache manager will always load and return the object, but it may evict it right away (Refer to [section](https://guava.dev/releases/18.0/api/docs/com/google/common/cache/CacheBuilder.html#maximumWeight(long)) about setting maximum weight to 0). 

In this case, an allocation will be loaded and returned, but may be evicted before anything can be done with it. In order to free the memory, the write lock will need to be obtained. That is why in line 142 we obtain the readlock and then check if the allocation has already been closed.